added reconnect logic

weiyu2001 opened the pull request:
This is intended to fix the issue when the nfd is rebooted, the existing logic will have a broken pipe for ever. The fix is the anytime a write failed a reconnect task is scheduled on the same thread pool with a configurable 5 sec delay. The same task will also be rescheduled if the the connect failed whether it is the original connect or called from the subsequent reconnects. The write semaphore will not be released until connect is succesfully completed to avoid multiple outstanding scheduled reconnects.

jefft0 commented on src/net/named_data/jndn/transport/AsyncTcpTransport.java:
```diff
@@ -83,8 +83,8 @@ public void completed(Integer bytesRead, ByteBuffer data) {
       public void failed(Throwable ex, ByteBuffer data) {
-        writeLock_.release();
         logger_.log(Level.SEVERE, null, ex);
+        scheduleReconnect();
```
Hi thanks for the pull request. Here you call scheduleReconnect if writeCompletionHandler_. But you don't call it on line 66 when readCompletionHandler_ fails. I wonder, if the connection drops while waiting for an incoming packet (without writing to the socket), will Java call readCompletionHandler_.failed?

weiyu2001 replied:
For broken pipe the readCompletionHandler_.failed is not called.

jefft0 replied:
OK, so we only find out about a dropped connection on write. That's OK for starters.

weiyu2001 replied:
I should qualify the above comment that we only tested the use cases when it is used in the face. Maybe need more tests standalone

jefft0 replied:
For a dropped connection on write, does it keep trying to reconnect forever? I don't see a limit count.

weiyu2001 replied:
No limit now. it will keep retrying until nfd is back up again. We try to avoid back to back busy immediate retry by introducing 5 sec delay. We also considered introducing the exponential backoff to lengthen the delay for each retry but realized the client will probably expect the prompt reconnect after nfd is back up after a relative long time down time. Any suggestions?

weiyu2001 replied:
Another interesting side effect of the current implementation is that for each new send() we first check if the connection is good getIsConnected() before we tried to lock the semaphore. It turned out that channel_.getRemoteAddress() throws connection refused IOException so that send() will not have to wait the default lock time 10 sec. The lock is not released if the reconnection keeps failing.

andrewsbrown replied:
I have a comment about the `readCompletionHandler_.failed()` comments above: like @weiyu2001 says, it probably bears more testing. When he tested (correct me here if I'm wrong), they restarted the NFD while simultaneously trying to `expressInterest()` in a loop. This test would not find a case when a read fails midway through because the client isn't reading. A better test might `registerPrefix()` and have some incoming interests coming in but it will be very difficult (if not impossible) to reliably cut the pipe in the middle of one of these reads. So I think we may want to consider adding the reconnection logic if the read fails, although I'm not sure how that will work with locking (e.g. if multiple reads fail in a row they could spawn a bunch of reconnect tasks).

jefft0 commented:
Would to be willing to add a flag to ConnectionInfo (default false) to enable the feature? Maybe:

    /**
     * Create a ConnectionInfo with the given host and port.
     * @param host The host for the connection.
     * @param port The port number for the connection.
     * @param blockForReconnect If true, then block on send until a dropped connection is reconnected. If false, send throws an exception if the connection is dropped.
     */
    public
    ConnectionInfo(String host, int port, boolean blockForReconnect)
    {
      host_ = host;
      port_ = port;
      blockForReconnect_ = blockForReconnect;
    }

weiyu2001 commented:
Jeff/Andrew, I added scheduleReconnect() in the readCompletionHandler_.failed() with reconnectLock_ to make sure only one reconnect is scheduled until reconnect is successful, I also added blockForReconnect option in ConnectionInfo. I didn't change any client so by default this is false.

jefft0 commented:
Looks good to me. Andrew, what do you think? Ready to merge?

andrewsbrown commented:
Functionality-wise, I think so; in our version @weiyu2001 added some comments about what happens when that may be helpful. Do you want those?

jefft0 commented:
Yes, please. Let me know when you add the comments and we can merge.

weiyu2001 commented:
added comments and logging

weiyu2001 commented:
@jefft0 Don't merge yet, one of our engineers had some valid inputs that I wanted to incorporate

jefft0 commented:
OK, I'll wait.

jefft0 commented on src/net/named_data/jndn/transport/AsyncTcpTransport.java:
```diff
@@ -39,6 +39,27 @@
  * AsyncTcpTransport extends Transport for async communication over TCP by
  * dispatching reads from an AsynchronousSocketChannel to a
  * ScheduledExecutorService.
+ * There is reconnect to nfd logic implemented if any one of the following case happens:
```
Thanks for the detailed comments. Before "There is a reconnect", I suggest adding "If blockForReconnect in the ConnectionInfo is true, then ..."

weiyu2001 replied:
I hope you didn't misinterpret the intention of the flag. It is not really block for reconnect. When this flag is set, each send will try to acquire the write semaphore first and thus potentially can wait for the previous write to complete or fail. If it is not set then the first thing the send is going to do is check if the channel is still connected but this check sometimes throws exception when the underlying channel is gone thus bypass the wait and immediately return to the caller.

jefft0 commented on src/net/named_data/jndn/transport/AsyncTcpTransport.java:
```diff
@@ -97,26 +119,41 @@ public void failed(Throwable ex, ByteBuffer data) {
+     * @param host The host for the connection.
+     */
+    public
+    ConnectionInfo(String host)
```
Can you add the constructor for ConnectionInfo(String host, int port) ?  (It was there previously.)

weiyu2001 replied:
added in the latest commit

weiyu2001 commented:
In the last commit I also added a three test cases, since it has to deal with timing of reconnect it will take about 2 min to complete

andrewsbrown commented on src/net/named_data/jndn/transport/AsyncTcpTransport.java:
```diff
@@ -97,26 +120,54 @@ public void failed(Throwable ex, ByteBuffer data) {
+      host_ = host;
+      port_ = port;
+      blockForReconnect_ = false;
+    }
```
Could use `this(...)`

andrewsbrown commented on src/net/named_data/jndn/transport/AsyncTcpTransport.java:
```diff
@@ -187,8 +246,15 @@ public void failed(Throwable ex, ByteBuffer data) {
 
-    channel_ = AsynchronousSocketChannel.open
-      (AsynchronousChannelGroup.withThreadPool(threadPool_));
+	channelGroup_ = AsynchronousChannelGroup.withThreadPool(threadPool_);
```
Spacing... @jefft0, will you run a formatter over this as a final commit?

jefft0 replied:
I usually do formatting and grammar fixes after merging the pull request.

andrewsbrown commented on tests/src/net/named_data/jndn/tests/unit_tests/AsyncTcpTransportTest.java:
```diff
@@ -0,0 +1,248 @@
+	}
+
+	private void stopServer() {
+		try {
```
I don't think this works correctly; when I run this test I can't start a new server because of:

```
java.net.BindException: Address already in use: JVM_Bind
	at java.net.TwoStacksPlainSocketImpl.socketBind(Native Method)
	at java.net.TwoStacksPlainSocketImpl.socketBind(TwoStacksPlainSocketImpl.java:137)
	at java.net.AbstractPlainSocketImpl.bind(AbstractPlainSocketImpl.java:387)
	at java.net.TwoStacksPlainSocketImpl.bind(TwoStacksPlainSocketImpl.java:110)
	at java.net.PlainSocketImpl.bind(PlainSocketImpl.java:190)
	at java.net.ServerSocket.bind(ServerSocket.java:375)
	at java.net.ServerSocket.<init>(ServerSocket.java:237)
	at java.net.ServerSocket.<init>(ServerSocket.java:128)
	at net.named_data.jndn.tests.unit_tests.AsyncTcpTransportTest$1.run(AsyncTcpTransportTest.java:215)
	at java.lang.Thread.run(Thread.java:745)
```

andrewsbrown replied:
So one test passes and the other two fail.

weiyu2001 replied:
Hmm, yesterday I spent some time to make sure each test properly shutdown the server. Did you run on Linux or Windows?

andrewsbrown replied:
Windows

weiyu2001 replied:
Hmm, just cloned onto windows and mvn install build and tests all pass including that newly added one. It's probably timing issue, one solution would be to use different port for each test

andrewsbrown commented on tests/src/net/named_data/jndn/tests/unit_tests/AsyncTcpTransportTest.java:
```diff
@@ -0,0 +1,248 @@
+			transport.send(ByteBuffer.wrap("exit\n".getBytes()));
+		} catch (IOException e) {
+			// TODO Auto-generated catch block
+			e.printStackTrace();
```
We should use the library's log utilities here

weiyu2001 replied:
logger utility is used. Many of the exception logged are from AsyncTcpTransport

andrewsbrown commented on tests/src/net/named_data/jndn/tests/unit_tests/AsyncTcpTransportTest.java:
```diff
@@ -0,0 +1,248 @@
+				ServerSocket server = null;
+				try {
+					server = new ServerSocket(port);
+					System.out.println("Server socket created");
```
We should use the library's log utilities here

jefft0 commented on tests/src/net/named_data/jndn/tests/unit_tests/AsyncTcpTransportTest.java:
```diff
@@ -0,0 +1,248 @@
+
+		// allow the server enough time to collect the pings
+		sleep(1000);
+		assertEquals(5, count);
```
For the continuous integration workflow, I'm a little worried about a test which takes two minutes to run. Also, there aren't a lot of "assert" statements, and maybe the test would be easier to write if it doesn't have to confirm to JUnit. Could this test be moved to the examples folder, like we did with Andrew's TestRemotePrefixRegistration? https://github.com/named-data/jndn/blob/master/examples/src/net/named_data/jndn/tests/TestRemotePrefixRegistration.java

weiyu2001 replied:
Totally agree, it is not really "unit" test, just some code to verify that reconnect actually happen under various scenarios

andrewsbrown commented:
All concerns addressed, closing and continuing in rebased #13
